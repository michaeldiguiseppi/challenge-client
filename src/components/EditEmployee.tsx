import React, { useRef, useState, useEffect } from "react";
import { gql, useMutation } from "@apollo/client";
import { useHistory } from "react-router-dom";
import type { Employee } from "./types";
import "./stylesheets/EditAndDisplayEmployee.css";

export const EDIT_EMPLOYEE = gql`
  mutation EditPerson($email: String!, $payload: EditPerson) {
    editPerson(email: $email, payload: $payload) {
      name {
        first
        last
        title
      }
      email
      picture {
        thumbnail
        large
        medium
      }
    }
  }
`;

export enum Titles {
  MR = "Mr",
  MS = "Ms",
  MRS = "Mrs",
  MISS = "Miss",
  MONSIEUR = "Monsieur",
  MADAME = "Madame",
}

export type EditEmployeeProps = {
  editFinished: (finished: boolean) => void;
  employee: Employee;
};

const EditEmployee: React.FC<EditEmployeeProps> = ({
  editFinished,
  employee,
}) => {
  const history = useHistory();
  const [editEmployee, { data }] = useMutation(EDIT_EMPLOYEE);

  useEffect(() => {
    if (data) {
      // Only replace the route if the email changes
      if (data.editPerson.email !== employee.email) {
        history.replace(`/people/${encodeURIComponent(data.editPerson.email)}`);
      }
      editFinished(false);
    }
  }, [data, editFinished, history, employee]);

  //State
  const [first, setFirst] = useState(employee?.name.first);
  const [last, setLast] = useState(employee?.name.last);
  const [email, setEmail] = useState(employee?.email);
  const [title, setTitle] = useState(employee?.name.title);

  //Refs
  const firstNameRef = useRef<HTMLInputElement>(null);
  const lastNameRef = useRef<HTMLInputElement>(null);
  const titleRef = useRef<HTMLSelectElement>(null);
  const emailRef = useRef<HTMLInputElement>(null);

  return (
    <div className="employee-edit-detail">
      <form
        onSubmit={e => {
          e.preventDefault();
          editEmployee({
            variables: {
              email: employee.email,
              payload: {
                title,
                first,
                last,
                email,
              },
            },
          });
        }}
      >
        <div className="employee-edit-header-photo">
          <img src={employee.picture?.large} alt={`${first}-${last}`} />
        </div>
        <div className="employee-edit-header-details">
          <div className="employee-edit-header-name">
            <div className="employee-edit-header-title">
              <div className="label-holder">
                <label htmlFor="titleField">Title: </label>
              </div>
              <div className="input-holder">
                <select
                  ref={titleRef}
                  defaultValue={title}
                  name="titleField"
                  onChange={e => setTitle(e.target.value)}
                >
                  {Object.values(Titles).map(title => (
                    <option value={title} key={title}>
                      {title}
                    </option>
                  ))}
                </select>
              </div>
            </div>
            <div className="employee-edit-header-first">
              <div className="label-holder">
                <label htmlFor="firstNameField">First: </label>
              </div>
              <div className="input-holder">
                <input
                  ref={firstNameRef}
                  value={first}
                  type="text"
                  name="firstNameField"
                  onChange={e => setFirst(e.target.value)}
                  data-testid="edit-employee-first-name"
                />
              </div>
            </div>
            <div className="employee-edit-header-last">
              <div className="label-holder">
                <label htmlFor="lastNameField">Last: </label>
              </div>
              <div className="input-holder">
                <input
                  ref={lastNameRef}
                  value={last}
                  type="text"
                  name="lastNameField"
                  onChange={e => setLast(e.target.value)}
                  data-testid="edit-employee-last-name"
                />
              </div>
            </div>
            <div className="employee-edit-header-email">
              <div className="label-holder">
                <label htmlFor="emailField">Email: </label>
              </div>
              <div className="input-holder">
                <input
                  type="email"
                  ref={emailRef}
                  value={email}
                  name="emailField"
                  onChange={e => setEmail(e.target.value)}
                  data-testid="edit-employee-email"
                />
              </div>
            </div>
            <div className="save-button-container">
              <button type="submit" className="save-button">
                Save
              </button>
            </div>
          </div>
        </div>
      </form>
    </div>
  );
};

export default EditEmployee;
