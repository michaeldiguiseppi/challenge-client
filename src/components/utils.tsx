import type { Employee } from "./types";

export const mockEmployees: Employee[] = [
  {
    name: {
      first: "Niclas",
      last: "Stornes",
      title: "Mr",
    },
    email: "niclas.stornes@example.com",
    picture: {
      thumbnail: "https://randomuser.me/api/portraits/thumb/men/78.jpg",
    },
    id: null,
  },
  {
    name: {
      first: "Erik",
      last: "Banks",
      title: "Mr",
    },
    email: "erik.banks@example.com",
    picture: {
      thumbnail: "https://randomuser.me/api/portraits/thumb/men/67.jpg",
      large: "https://randomuser.me/api/portraits/men/67.jpg",
      medium: "https://randomuser.me/api/portraits/med/men/67.jpg",
    },
    id: null,
  },
  {
    name: {
      first: "Andrea",
      last: "Crespo",
      title: "Mrs",
    },
    email: "andrea.crespo@example.com",
    picture: {
      thumbnail: "https://randomuser.me/api/portraits/thumb/women/41.jpg",
      large: "https://randomuser.me/api/portraits/women/41.jpg",
      medium: "https://randomuser.me/api/portraits/med/women/41.jpg",
    },
    id: null,
  },
];
