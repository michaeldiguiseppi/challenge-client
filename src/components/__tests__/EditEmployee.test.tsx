import React from "react";
import { render, cleanup, waitFor, fireEvent } from "@testing-library/react";
import EditEmployee, { EDIT_EMPLOYEE } from "../EditEmployee";
import type { EditEmployeeProps } from "../EditEmployee";
import type { Employee } from "../types";
import { mockEmployees } from "../utils";
import { MockedProvider } from "@apollo/client/testing";

const mocks = [
  {
    request: {
      query: EDIT_EMPLOYEE,
      variables: {
        email: mockEmployees[0].email,
        payload: {
          title: mockEmployees[0].name.title,
          first: mockEmployees[0].name.first,
          last: mockEmployees[0].name.last,
          email: mockEmployees[0].email,
        },
      },
    },
    newData: jest.fn(() => ({
      data: {
        person: {
          name: {
            first: "Niclas",
            last: "Stornes",
            title: "Mr",
          },
          email: "niclas.stornes@example.com",
          picture: {
            thumbnail: "https://randomuser.me/api/portraits/thumb/men/78.jpg",
            large: "https://randomuser.me/api/portraits/large/men/78.jpg",
            medium: "https://randomuser.me/api/portraits/medium/men/78.jpg",
          },
        },
      },
    })),
  },
];

beforeEach(cleanup);

const employee: Employee = mockEmployees[0];

const Props: EditEmployeeProps = {
  employee,
  editFinished: jest.fn(),
};

describe("<EditEmployee />", () => {
  it("renders without crashing", async () => {
    const { container } = render(
      <MockedProvider mocks={mocks}>
        <EditEmployee {...Props} />
      </MockedProvider>
    );
    expect(container).toMatchSnapshot();
  });

  it("should render a large profile image for the selected employee", () => {
    const { getByAltText } = render(
      <MockedProvider mocks={mocks}>
        <EditEmployee {...Props} />
      </MockedProvider>
    );
    getByAltText(/niclas-stornes/i);
  });

  it("should render an input field for the first name", () => {
    const { getByTestId } = render(
      <MockedProvider mocks={mocks}>
        <EditEmployee {...Props} />
      </MockedProvider>
    );
    getByTestId(/edit-employee-first-name/i);
  });

  it("should render an input field for the last name", () => {
    const { getByTestId } = render(
      <MockedProvider mocks={mocks}>
        <EditEmployee {...Props} />
      </MockedProvider>
    );
    getByTestId(/edit-employee-last-name/i);
  });

  it("should render an input field for the email", () => {
    const { getByTestId } = render(
      <MockedProvider mocks={mocks}>
        <EditEmployee {...Props} />
      </MockedProvider>
    );
    getByTestId(/edit-employee-email/i);
  });

  it("should render a save button", () => {
    const { getByText } = render(
      <MockedProvider mocks={mocks}>
        <EditEmployee {...Props} />
      </MockedProvider>
    );
    getByText(/save/i);
  });

  it("should call the edit employee mutation when the save button is pressed", async () => {
    const { getByText } = render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <EditEmployee {...Props} />
      </MockedProvider>
    );
    const saveButton = getByText(/save/i);
    fireEvent.click(saveButton);
    const editEmployeeMutationMock = mocks[0].newData;
    await waitFor(() => expect(editEmployeeMutationMock).toHaveBeenCalled());
  });
});
