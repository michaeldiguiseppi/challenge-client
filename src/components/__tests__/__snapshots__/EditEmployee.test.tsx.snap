// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EditEmployee /> renders without crashing 1`] = `
<div>
  <div
    class="employee-edit-detail"
  >
    <form>
      <div
        class="employee-edit-header-photo"
      >
        <img
          alt="Niclas-Stornes"
        />
      </div>
      <div
        class="employee-edit-header-details"
      >
        <div
          class="employee-edit-header-name"
        >
          <div
            class="employee-edit-header-title"
          >
            <div
              class="label-holder"
            >
              <label
                for="titleField"
              >
                Title: 
              </label>
            </div>
            <div
              class="input-holder"
            >
              <select
                name="titleField"
              >
                <option
                  selected=""
                  value="Mr"
                >
                  Mr
                </option>
                <option
                  value="Ms"
                >
                  Ms
                </option>
                <option
                  value="Mrs"
                >
                  Mrs
                </option>
                <option
                  value="Miss"
                >
                  Miss
                </option>
                <option
                  value="Monsieur"
                >
                  Monsieur
                </option>
                <option
                  value="Madame"
                >
                  Madame
                </option>
              </select>
            </div>
          </div>
          <div
            class="employee-edit-header-first"
          >
            <div
              class="label-holder"
            >
              <label
                for="firstNameField"
              >
                First: 
              </label>
            </div>
            <div
              class="input-holder"
            >
              <input
                data-testid="edit-employee-first-name"
                name="firstNameField"
                value="Niclas"
              />
            </div>
          </div>
          <div
            class="employee-edit-header-last"
          >
            <div
              class="label-holder"
            >
              <label
                for="lastNameField"
              >
                Last: 
              </label>
            </div>
            <div
              class="input-holder"
            >
              <input
                data-testid="edit-employee-last-name"
                name="lastNameField"
                value="Stornes"
              />
            </div>
          </div>
          <div
            class="employee-edit-header-email"
          >
            <div
              class="label-holder"
            >
              <label
                for="emailField"
              >
                Email: 
              </label>
            </div>
            <div
              class="input-holder"
            >
              <input
                data-testid="edit-employee-email"
                name="emailField"
                value="niclas.stornes@example.com"
              />
            </div>
          </div>
          <div
            class="save-button-container"
          >
            <button
              class="save-button"
              type="submit"
            >
              Save
            </button>
          </div>
        </div>
      </div>
    </form>
  </div>
</div>
`;
